File: kotlin/src/test/kotlin/com/bhengubv/circleai/IdentityTest.kt
// IdentityTest.kt
//
// Verifies IdentityTier enum values and CircleIdentity / RegisteredDevice
// data class construction against the three fixture examples in identity.json.

package com.bhengubv.circleai

import com.bhengubv.circleai.identity.CircleIdentity
import com.bhengubv.circleai.identity.IdentityTier
import com.bhengubv.circleai.identity.RegisteredDevice
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.jupiter.api.Test
import java.io.File
import java.time.Instant
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class IdentityTest {

    private val json = Json { ignoreUnknownKeys = true }

    private fun locateFixture(name: String): File {
        // Walk UP from the working directory looking for a `fixtures` dir.
        // The previous version hardcoded a Windows absolute path with a
        // relative fallback that resolved on neither CI nor a Mac, so these
        // tests could only pass on one machine.
        var dir: File? = File(".").absoluteFile
        while (dir != null) {
            val candidate = File(dir, "fixtures/" + name)
            if (candidate.exists()) return candidate
            dir = dir.parentFile
        }
        error("Cannot locate fixture " + name)
    }

    // ── IdentityTier enum ─────────────────────────────────────────────────────

    @Test
    fun `IdentityTier has exactly 3 values`() {
        assertEquals(3, IdentityTier.entries.size)
    }

    @Test
    fun `IdentityTier values are Anonymous Pseudonymous Verified in order`() {
        val values = IdentityTier.entries.map { it.name }
        assertEquals(listOf("Anonymous", "Pseudonymous", "Verified"), values)
    }

    @Test
    fun `IdentityTier ordinals are correct`() {
        assertEquals(0, IdentityTier.Anonymous.ordinal)
        assertEquals(1, IdentityTier.Pseudonymous.ordinal)
        assertEquals(2, IdentityTier.Verified.ordinal)
    }

    // ── Fixture: verified_multi_device ────────────────────────────────────────

    @Test
    fun `fixture verified_multi_device identity parses correctly`() {
        val root = json.parseToJsonElement(locateFixture("identity.json").readText()).jsonObject
        val example = root["examples"]!!.jsonArray.first { it.jsonObject["id"]!!.jsonPrimitive.content == "verified_multi_device" }.jsonObject
        val id = example["identity"]!!.jsonObject

        val identity = CircleIdentity(
            identityId        = id["identityId"]!!.jsonPrimitive.content,
            displayName       = id["displayName"]!!.jsonPrimitive.content,
            preferredLanguage = id["preferredLanguage"]!!.takeIf { it !is JsonNull }?.jsonPrimitive?.content,
            tier              = IdentityTier.valueOf(id["tier"]!!.jsonPrimitive.content),
            deviceIds         = id["deviceIds"]!!.jsonArray.map { it.jsonPrimitive.content },
            createdAt         = Instant.parse(id["createdAt"]!!.jsonPrimitive.content),
            lastSeenAt        = Instant.parse(id["lastSeenAt"]!!.jsonPrimitive.content)
        )

        assertEquals("a1b2c3d4-e5f6-7890-abcd-ef1234567890", identity.identityId)
        assertEquals("Sipho Dlamini", identity.displayName)
        assertEquals("zu", identity.preferredLanguage)
        assertEquals(IdentityTier.Verified, identity.tier)
        assertEquals(3, identity.deviceIds.size)
        assertEquals("d1000000-0000-0000-0000-000000000001", identity.deviceIds[0])
        assertEquals("d3000000-0000-0000-0000-000000000003", identity.deviceIds[2])
        assertEquals(Instant.parse("2025-01-15T08:00:00Z"), identity.createdAt)
        assertEquals(Instant.parse("2026-05-12T09:30:00Z"), identity.lastSeenAt)
    }

    @Test
    fun `fixture verified_multi_device devices parse correctly`() {
        val root = json.parseToJsonElement(locateFixture("identity.json").readText()).jsonObject
        val example = root["examples"]!!.jsonArray.first { it.jsonObject["id"]!!.jsonPrimitive.content == "verified_multi_device" }.jsonObject
        val devicesArray = example["devices"]!!.jsonArray

        val devices = devicesArray.map { element ->
            val d = element.jsonObject
            RegisteredDevice(
                deviceId     = d["deviceId"]!!.jsonPrimitive.content,
                identityId   = d["identityId"]!!.jsonPrimitive.content,
                platform     = d["platform"]!!.jsonPrimitive.content,
                deviceName   = d["deviceName"]!!.takeIf { it !is JsonNull }?.jsonPrimitive?.content,
                registeredAt = Instant.parse(d["registeredAt"]!!.jsonPrimitive.content),
                lastActiveAt = Instant.parse(d["lastActiveAt"]!!.jsonPrimitive.content)
            )
        }

        assertEquals(3, devices.size)
        assertEquals("android",  devices[0].platform)
        assertEquals("Samsung Galaxy S25", devices[0].deviceName)
        assertEquals("watch",    devices[1].platform)
        assertEquals("Galaxy Watch 7", devices[1].deviceName)
        assertEquals("windows",  devices[2].platform)
        assertEquals("Work Laptop", devices[2].deviceName)
    }

    // ── Fixture: pseudonymous_single_device ───────────────────────────────────

    @Test
    fun `fixture pseudonymous_single_device identity parses correctly`() {
        val root = json.parseToJsonElement(locateFixture("identity.json").readText()).jsonObject
        val example = root["examples"]!!.jsonArray.first { it.jsonObject["id"]!!.jsonPrimitive.content == "pseudonymous_single_device" }.jsonObject
        val id = example["identity"]!!.jsonObject

        val identity = CircleIdentity(
            identityId        = id["identityId"]!!.jsonPrimitive.content,
            displayName       = id["displayName"]!!.jsonPrimitive.content,
            preferredLanguage = id["preferredLanguage"]!!.takeIf { it !is JsonNull }?.jsonPrimitive?.content,
            tier              = IdentityTier.valueOf(id["tier"]!!.jsonPrimitive.content),
            deviceIds         = id["deviceIds"]!!.jsonArray.map { it.jsonPrimitive.content },
            createdAt         = Instant.parse(id["createdAt"]!!.jsonPrimitive.content),
            lastSeenAt        = Instant.parse(id["lastSeenAt"]!!.jsonPrimitive.content)
        )

        assertEquals(IdentityTier.Pseudonymous, identity.tier)
        assertEquals("B! User", identity.displayName)
        assertEquals("en", identity.preferredLanguage)
        assertEquals(1, identity.deviceIds.size)
    }

    // ── Fixture: anonymous_iot ────────────────────────────────────────────────

    @Test
    fun `fixture anonymous_iot has null preferredLanguage and null deviceName`() {
        val root = json.parseToJsonElement(locateFixture("identity.json").readText()).jsonObject
        val example = root["examples"]!!.jsonArray.first { it.jsonObject["id"]!!.jsonPrimitive.content == "anonymous_iot" }.jsonObject
        val id = example["identity"]!!.jsonObject

        val identity = CircleIdentity(
            identityId        = id["identityId"]!!.jsonPrimitive.content,
            displayName       = id["displayName"]!!.jsonPrimitive.content,
            preferredLanguage = id["preferredLanguage"]!!.takeIf { it !is JsonNull }?.jsonPrimitive?.content,
            tier              = IdentityTier.valueOf(id["tier"]!!.jsonPrimitive.content),
            deviceIds         = id["deviceIds"]!!.jsonArray.map { it.jsonPrimitive.content },
            createdAt         = Instant.parse(id["createdAt"]!!.jsonPrimitive.content),
            lastSeenAt        = Instant.parse(id["lastSeenAt"]!!.jsonPrimitive.content)
        )

        assertEquals(IdentityTier.Anonymous, identity.tier)
        assertEquals("Guest", identity.displayName)
        assertNull(identity.preferredLanguage)

        val deviceNode = example["devices"]!!.jsonArray[0].jsonObject
        val device = RegisteredDevice(
            deviceId     = deviceNode["deviceId"]!!.jsonPrimitive.content,
            identityId   = deviceNode["identityId"]!!.jsonPrimitive.content,
            platform     = deviceNode["platform"]!!.jsonPrimitive.content,
            deviceName   = deviceNode["deviceName"]!!.takeIf { it !is JsonNull }?.jsonPrimitive?.content,
            registeredAt = Instant.parse(deviceNode["registeredAt"]!!.jsonPrimitive.content),
            lastActiveAt = Instant.parse(deviceNode["lastActiveAt"]!!.jsonPrimitive.content)
        )

        assertEquals("iot", device.platform)
        assertNull(device.deviceName)
    }

    // ── Platforms list ────────────────────────────────────────────────────────

    @Test
    fun `fixture platforms list has 8 entries`() {
        val root = json.parseToJsonElement(locateFixture("identity.json").readText()).jsonObject
        val platforms = root["platforms"]!!.jsonArray.map { it.jsonPrimitive.content }
        assertEquals(8, platforms.size)
        assertTrue(platforms.containsAll(listOf("android", "ios", "windows", "macos", "linux", "web", "watch", "iot")))
    }

    // ── Data class structural checks ──────────────────────────────────────────

    @Test
    fun `CircleIdentity copy works correctly`() {
        val identity = CircleIdentity(
            identityId        = "test-id",
            displayName       = "Test User",
            preferredLanguage = "en",
            tier              = IdentityTier.Verified,
            deviceIds         = listOf("device-1"),
            createdAt         = Instant.EPOCH,
            lastSeenAt        = Instant.EPOCH
        )
        val updated = identity.copy(displayName = "Updated User")
        assertEquals("Updated User", updated.displayName)
        assertEquals("test-id", updated.identityId)
    }
}

File: harmonyos/tests/identity.test.ts
// identity.test.ts
//
// Verifies IdentityTier enum, CircleIdentity schema, RegisteredDevice schema,
// and fixture examples from fixtures/identity.json — HarmonyOS/ArkTS port.

import * as fs from 'fs';
import * as path from 'path';
import { IdentityTier, type CircleIdentity, type RegisteredDevice } from '../src/identity';

// ---------------------------------------------------------------------------
// Fixture types
// ---------------------------------------------------------------------------

interface IdentityFixtureExample {
  id:          string;
  description: string;
  identity: {
    identityId:        string;
    displayName:       string;
    preferredLanguage: string | null;
    tier:              string;
    deviceIds:         string[];
    createdAt:         string;
    lastSeenAt:        string;
  };
  devices: Array<{
    deviceId:     string;
    identityId:   string;
    platform:     string;
    deviceName:   string | null;
    registeredAt: string;
    lastActiveAt: string;
  }>;
}

interface IdentityFixture {
  identityTiers: string[];
  platforms:     string[];
  examples:      IdentityFixtureExample[];
  assertions: {
    tierOrder:       string[];
    identityIdFormat: string;
    deviceIdFormat:   string;
    timestampFormat:  string;
  };
}

// ---------------------------------------------------------------------------
// Load fixture
// ---------------------------------------------------------------------------

const fixturePath = path.resolve(__dirname, '../../fixtures/identity.json');
const fixture: IdentityFixture = JSON.parse(fs.readFileSync(fixturePath, 'utf-8'));

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------

function buildIdentity(ex: IdentityFixtureExample): CircleIdentity {
  return {
    identityId:        ex.identity.identityId,
    displayName:       ex.identity.displayName,
    preferredLanguage: ex.identity.preferredLanguage,
    tier:              ex.identity.tier as IdentityTier,
    deviceIds:         ex.identity.deviceIds,
    createdAt:         new Date(ex.identity.createdAt),
    lastSeenAt:        new Date(ex.identity.lastSeenAt),
  };
}

function buildDevices(ex: IdentityFixtureExample): RegisteredDevice[] {
  return ex.devices.map(d => ({
    deviceId:     d.deviceId,
    identityId:   d.identityId,
    platform:     d.platform,
    deviceName:   d.deviceName,
    registeredAt: new Date(d.registeredAt),
    lastActiveAt: new Date(d.lastActiveAt),
  }));
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe('IdentityTier enum', () => {
  test('has three values', () => {
    const tiers = Object.values(IdentityTier);
    expect(tiers.length).toBe(3);
  });

  test('contains Anonymous', () => {
    expect(IdentityTier.Anonymous).toBe('Anonymous');
  });

  test('contains Pseudonymous', () => {
    expect(IdentityTier.Pseudonymous).toBe('Pseudonymous');
  });

  test('contains Verified', () => {
    expect(IdentityTier.Verified).toBe('Verified');
  });

  test('tier values match fixture tierOrder', () => {
    for (const tier of fixture.assertions.tierOrder) {
      expect(Object.values(IdentityTier)).toContain(tier);
    }
  });

  test('all fixture identity tiers are valid IdentityTier values', () => {
    for (const tier of fixture.identityTiers) {
      expect(Object.values(IdentityTier)).toContain(tier);
    }
  });
});

describe('CircleIdentity schema', () => {
  test.each(fixture.examples)('$id — $description', (ex) => {
    const identity = buildIdentity(ex);

    expect(typeof identity.identityId).toBe('string');
    expect(identity.identityId.length).toBeGreaterThan(0);
    expect(typeof identity.displayName).toBe('string');
    expect(Object.values(IdentityTier)).toContain(identity.tier);
    expect(Array.isArray(identity.deviceIds)).toBe(true);
    expect(identity.createdAt instanceof Date).toBe(true);
    expect(identity.lastSeenAt instanceof Date).toBe(true);
    expect(identity.lastSeenAt >= identity.createdAt).toBe(true);
  });

  test('verified_multi_device has 3 device IDs', () => {
    const ex = fixture.examples.find(e => e.id === 'verified_multi_device')!;
    const identity = buildIdentity(ex);
    expect(identity.deviceIds.length).toBe(3);
    expect(identity.tier).toBe(IdentityTier.Verified);
  });

  test('pseudonymous_single_device has Pseudonymous tier', () => {
    const ex = fixture.examples.find(e => e.id === 'pseudonymous_single_device')!;
    const identity = buildIdentity(ex);
    expect(identity.tier).toBe(IdentityTier.Pseudonymous);
    expect(identity.preferredLanguage).toBe('en');
  });

  test('anonymous_iot has null preferredLanguage', () => {
    const ex = fixture.examples.find(e => e.id === 'anonymous_iot')!;
    const identity = buildIdentity(ex);
    expect(identity.tier).toBe(IdentityTier.Anonymous);
    expect(identity.preferredLanguage).toBeNull();
  });
});

describe('RegisteredDevice schema', () => {
  test('all fixture platforms are known platform strings', () => {
    const knownPlatforms = new Set(fixture.platforms);
    for (const ex of fixture.examples) {
      for (const d of buildDevices(ex)) {
        expect(knownPlatforms.has(d.platform)).toBe(true);
      }
    }
  });

  test('all devices have matching identityId', () => {
    for (const ex of fixture.examples) {
      const devices = buildDevices(ex);
      for (const d of devices) {
        expect(d.identityId).toBe(ex.identity.identityId);
      }
    }
  });

  test('anonymous_iot device has null deviceName', () => {
    const ex = fixture.examples.find(e => e.id === 'anonymous_iot')!;
    const devices = buildDevices(ex);
    expect(devices[0].deviceName).toBeNull();
    expect(devices[0].platform).toBe('iot');
  });

  test('verified_multi_device has android, watch, windows platforms', () => {
    const ex = fixture.examples.find(e => e.id === 'verified_multi_device')!;
    const platforms = buildDevices(ex).map(d => d.platform);
    expect(platforms).toContain('android');
    expect(platforms).toContain('watch');
    expect(platforms).toContain('windows');
  });
});
